src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -97,88 +93,103 @@ module IncrementalBuilderEventTesting =
-                    else
-                        use text = source.GetTextContainer()
-                        match text with
-                        | TextContainer.Stream(stream) ->
-                            ParseOneInputStream(tcConfig, lexResourceManager, fileName, isLastCompiland, diagnosticsLogger, false, stream)
-                        | TextContainer.SourceText(sourceText) ->
-                            ParseOneInputSourceText(tcConfig, lexResourceManager, fileName, isLastCompiland, diagnosticsLogger, sourceText)
-                        | TextContainer.OnDisk ->
-                            ParseOneInputFile(tcConfig, lexResourceManager, fileName, isLastCompiland, diagnosticsLogger, true)
+            try           
+                use text = source.GetTextContainer()
+                let input = 
+                    match text with
+                    | TextContainer.Stream(stream) ->
+                        ParseOneInputStream(tcConfig, lexResourceManager, fileName, isLastCompiland, diagnosticsLogger, false, stream)
+                    | TextContainer.SourceText(sourceText) ->
+                        ParseOneInputSourceText(tcConfig, lexResourceManager, fileName, isLastCompiland, diagnosticsLogger, sourceText)
+                    | TextContainer.OnDisk ->
+                        ParseOneInputFile(tcConfig, lexResourceManager, fileName, isLastCompiland, diagnosticsLogger, true)
 
                 fileParsed.Trigger fileName
 
-                let res = input, sourceRange, fileName, diagnosticsLogger.GetDiagnostics()
-                // If we do not skip parsing the file, then we can cache the real result.
-                if not canSkip then
-                    weakCache <- Some(WeakReference<_>(res))
-                res
+                input, sourceRange, fileName, diagnosticsLogger.GetDiagnostics()
+
             with exn ->
                 let msg = sprintf "unexpected failure in SyntaxTree.parse\nerror = %s" (exn.ToString())
                 System.Diagnostics.Debug.Assert(false, msg)
                 failwith msg
 
-        /// Parse the given file and return the given input.
-        member _.Parse sigNameOpt =
-            match weakCache with
-            | Some weakCache ->
-                match weakCache.TryGetTarget() with
-                | true, res -> res
-                | _ -> parse sigNameOpt
-            | _ -> parse sigNameOpt
+        let parseOrSkip sigNameOpt =
+            IncrementalBuilderEventTesting.MRU.Add(IncrementalBuilderEventTesting.IBEParsed fileName)
+            use _ =
+                Activity.start "IncrementalBuildSyntaxTree.parseOrSkip"
+                    [|
+                        Activity.Tags.fileName, fileName                       
+                        "buildPhase", BuildPhase.Parse.ToString()
+                    |]
+
+            match tryImplStubWhenHasSig sigNameOpt with
+            | Some result ->
+                Activity.addEvent "ImplementationSkipped"
```

Comment:
Could this be a tag on the activity above, rather than a separate event? Counting that would probably be quite simple still.

Reply:
Absolutely. Would it be better to add the tag conditionally or always (with value "false" when not skipping)?

Reply:
Done, I restored "canSkip". BTW after all this refactoring I can see there may be a low cost path to parallelizing parsing.